File: tools/zping.py
#!/usr/bin/env python3
"""zenoh bridge probe: round-trip latency (ping/pong) and throughput (flood/sink).

ping 端與 pong 端各跑一個。RTT 全程用 ping 端的時鐘量（pong 只是把 header.stamp
原封不動抄回 /goal_pose），所以兩台機器不需要對時。

topic 選 /initialpose、/scan、/d455/color/image_raw、/goal_pose 是因為
zenoh allowlist 只放行這幾個；量測期間這些 topic 在兩端都沒有其他訂閱者。

expect-none 角色是白名單閘門的否定測試：搭 --topic 指定一個「不該過橋」的
topic，收到任何一則就算失敗。白名單寫錯時橋不會報錯，只會靜默放行或靜默阻擋，
沒有這個角色就分不出「擋住了」與「根本沒在跑」。

expect-none 單獨跑沒有意義：橋掛掉、網路斷、domain 設錯時它一樣是綠燈。必須與
同一時窗、同一組行程的正向 ping（白名單內 topic 且必須收到）配成一對，正向那半
才是「鏈路真的在跑」的證據。配對由驗收腳本負責，不在本檔內。
"""
import argparse, json, sys, time

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from geometry_msgs.msg import PoseStamped, PoseWithCovarianceStamped
from sensor_msgs.msg import Image, LaserScan

MODES = {
    "small": (PoseWithCovarianceStamped, "/initialpose"),
    "scan":  (LaserScan,                 "/scan"),
    "image": (Image,                     "/d455/color/image_raw"),
}
REPLY_TOPIC = "/goal_pose"


def resolve(mode, topic):
    cls, default_topic = MODES[mode]
    return cls, (topic or default_topic)


def make_payload(mode, stamp, points, width, height):
    cls, _ = MODES[mode]
    msg = cls()
    msg.header.stamp = stamp
    msg.header.frame_id = "zping"
    if mode == "scan":
        msg.angle_min, msg.angle_max = -3.14, 3.14
        msg.angle_increment = 6.28 / points
        msg.range_min, msg.range_max = 0.1, 30.0
        msg.ranges = [1.0] * points
    elif mode == "image":
        msg.width, msg.height = width, height
        msg.encoding = "rgb8"
        msg.step = width * 3
        msg.data = bytes(width * height * 3)
    return msg


class Flood(Node):
    """單向灌流量：以固定速率發，序號放在 header.frame_id 讓 sink 算掉包。"""

    def __init__(self, mode, rate, seconds, reliable, points, width, height, topic=""):
        super().__init__("zflood")
        cls, topic = resolve(mode, topic)
        self.mode = mode
        self.points, self.width, self.height = points, width, height
        self.pub = self.create_publisher(cls, topic, qos(reliable))
        self.seq = 0
        self.deadline = time.time() + seconds
        self.create_timer(1.0 / rate, self.tick)

    def tick(self):
        if time.time() > self.deadline:
            return
        msg = make_payload(self.mode, self.get_clock().now().to_msg(),
                           self.points, self.width, self.height)
        msg.header.frame_id = str(self.seq)
        self.pub.publish(msg)
        self.seq += 1


class Sink(Node):
    """收端：算實際到達的訊息數、位元組數與序號缺口。"""

    def __init__(self, mode, reliable, topic=""):
        super().__init__("zsink")
        cls, topic = resolve(mode, topic)
        self.mode = mode
        self.sub = self.create_subscription(cls, topic, self.cb, qos(reliable))
        self.n, self.bytes = 0, 0
        self.first_t = None
        self.last_t = None
        self.seqs = []

    def cb(self, msg):
        now = time.time()
        if self.first_t is None:
            self.first_t = now
        self.last_t = now
        self.n += 1
        if self.mode == "image":
            self.bytes += len(msg.data)
        elif self.mode == "scan":
            self.bytes += 4 * len(msg.ranges)
        else:
            self.bytes += 736
        try:
            self.seqs.append(int(msg.header.frame_id))
        except ValueError:
            pass

    def report(self, label, offered_rate):
        dur = (self.last_t - self.first_t) if (self.n > 1) else 0.0
        out = {"label": label, "mode": self.mode, "offered_hz": offered_rate,
               "recv": self.n, "seconds": round(dur, 2)}
        if dur > 0:
            out["delivered_hz"] = round((self.n - 1) / dur, 1)
            out["delivered_MBps"] = round(self.bytes / dur / 1e6, 2)
        if self.seqs:
            sent_span = max(self.seqs) - min(self.seqs) + 1
            out["sent_span"] = sent_span
            out["loss_pct"] = round(100 * (1 - self.n / sent_span), 1)
        print("RESULT " + json.dumps(out))


class ExpectNone(Node):
    """否定測試：訂閱一個「不該送達」的 topic，收到任何一則即為失敗。

    刻意固定用 BEST_EFFORT 訂閱：BEST_EFFORT reader 同時相容 RELIABLE 與
    BEST_EFFORT writer，若跟著 --best-effort 走，QoS 不相容會讓訂閱根本配不上
    發布端，測試就會因為「沒配對」而假性通過。
    """

    def __init__(self, mode, topic=""):
        super().__init__("zexpect_none")
        cls, topic = resolve(mode, topic)
        self.mode_name, self.topic = mode, topic
        self.sub = self.create_subscription(cls, topic, self.cb, qos(False))
        self.n = 0

    def cb(self, msg):
        self.n += 1

    def report(self, label, seconds):
        ok = self.n == 0
        print("RESULT " + json.dumps({
            "label": label, "impl": "py", "role": "expect-none", "mode": self.mode_name,
            "topic": self.topic, "seconds": round(seconds, 2), "recv": self.n, "pass": ok}))
        print(f"PASS: no messages on {self.topic} in {seconds:.1f}s" if ok
              else f"FAIL: received {self.n} message(s) on {self.topic} in {seconds:.1f}s")
        return ok


def qos(reliable):
    return QoSProfile(
        reliability=ReliabilityPolicy.RELIABLE if reliable else ReliabilityPolicy.BEST_EFFORT,
        history=HistoryPolicy.KEEP_LAST, depth=10)


class Pong(Node):
    def __init__(self, mode, reliable, topic=""):
        super().__init__("zpong")
        cls, topic = resolve(mode, topic)
        self.pub = self.create_publisher(PoseStamped, REPLY_TOPIC, qos(True))
        self.sub = self.create_subscription(cls, topic, self.cb, qos(reliable))
        self.n = 0
        self.get_logger().info(f"pong: {topic} -> {REPLY_TOPIC}")

    def cb(self, msg):
        out = PoseStamped()
        out.header.stamp = msg.header.stamp      # 抄回原 stamp = ping 端時鐘
        out.header.frame_id = "zpong"
        self.pub.publish(out)
        self.n += 1
        if self.n % 20 == 0:
            self.get_logger().info(f"pong: {self.n} echoed")


class Ping(Node):
    def __init__(self, mode, count, rate, reliable, points, width, height, topic=""):
        super().__init__("zping")
        cls, topic = resolve(mode, topic)
        self.mode, self.count = mode, count
        self.points, self.width, self.height = points, width, height
        self.pub = self.create_publisher(cls, topic, qos(reliable))
        self.sub = self.create_subscription(PoseStamped, REPLY_TOPIC, self.cb, qos(True))
        self.rtts, self.sent = [], 0
        self.create_timer(1.0 / rate, self.tick)

    def tick(self):
        if self.sent >= self.count:
            return
        now = self.get_clock().now()
        self.pub.publish(make_payload(self.mode, now.to_msg(), self.points, self.width, self.height))
        self.sent += 1

    def cb(self, msg):
        sent_ns = msg.header.stamp.sec * 10**9 + msg.header.stamp.nanosec
        self.rtts.append((self.get_clock().now().nanoseconds - sent_ns) / 1e6)


def pct(v, p):
    s = sorted(v)
    return s[min(len(s) - 1, int(round(p / 100 * (len(s) - 1))))]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--role", choices=["ping", "pong", "flood", "sink", "expect-none"], required=True)
    ap.add_argument("--mode", choices=list(MODES), default="small")
    ap.add_argument("--topic", default="", help="覆寫 --mode 的預設 topic（訊息型別仍由 --mode 決定）")
    ap.add_argument("--count", type=int, default=100)
    ap.add_argument("--rate", type=float, default=10.0)
    ap.add_argument("--points", type=int, default=1080)
    ap.add_argument("--width", type=int, default=640)
    ap.add_argument("--height", type=int, default=480)
    ap.add_argument("--best-effort", action="store_true")
    ap.add_argument("--label", default="")
    ap.add_argument("--seconds", type=float, default=15.0)
    a = ap.parse_args()

    rclpy.init()
    reliable = not a.best_effort
    if a.role == "pong":
        node = Pong(a.mode, reliable, a.topic)
        try:
            rclpy.spin(node)
        except KeyboardInterrupt:
            pass
        rclpy.shutdown()
        return

    if a.role == "flood":
        node = Flood(a.mode, a.rate, a.seconds, reliable, a.points, a.width, a.height, a.topic)
        end = time.time() + a.seconds + 1.0
        while rclpy.ok() and time.time() < end:
            rclpy.spin_once(node, timeout_sec=0.05)
        print("RESULT " + json.dumps({"label": a.label, "role": "flood", "mode": a.mode,
                                      "offered_hz": a.rate, "published": node.seq,
                                      "actual_pub_hz": round(node.seq / a.seconds, 1)}))
        rclpy.shutdown()
        return

    if a.role == "expect-none":
        node = ExpectNone(a.mode, a.topic)
        end = time.time() + a.seconds
        while rclpy.ok() and time.time() < end:
            rclpy.spin_once(node, timeout_sec=0.05)
        ok = node.report(a.label, a.seconds)
        rclpy.shutdown()
        sys.exit(0 if ok else 1)

    if a.role == "sink":
        node = Sink(a.mode, reliable, a.topic)
        end = time.time() + a.seconds
        while rclpy.ok() and time.time() < end:
            rclpy.spin_once(node, timeout_sec=0.05)
        node.report(a.label, a.rate)
        rclpy.shutdown()
        return

    node = Ping(a.mode, a.count, a.rate, reliable, a.points, a.width, a.height, a.topic)
    deadline = time.time() + a.count / a.rate + 8.0
    while rclpy.ok() and time.time() < deadline and len(node.rtts) < a.count:
        rclpy.spin_once(node, timeout_sec=0.05)
    r = node.rtts
    out = {"label": a.label, "mode": a.mode, "sent": node.sent, "recv": len(r),
           "loss_pct": round(100 * (1 - len(r) / max(1, node.sent)), 1)}
    if r:
        out.update({"rtt_min_ms": round(min(r), 3), "rtt_p50_ms": round(pct(r, 50), 3),
                    "rtt_p90_ms": round(pct(r, 90), 3), "rtt_p99_ms": round(pct(r, 99), 3),
                    "rtt_max_ms": round(max(r), 3),
                    "owd_p50_ms": round(pct(r, 50) / 2, 3)})
    print("RESULT " + json.dumps(out))
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: tools/zping.cpp
// zping.py 的 rclcpp 版本：同樣的 topic、同樣的量法（pong 把 header.stamp 原樣抄回），
// 用來把「rclpy 的成本」從「橋 + 網路的成本」裡拆出來。
// expect-none 角色與 Python 版同義：斷言指定 topic 收不到任何訊息，收到即失敗（exit 1）。
// 它單獨跑沒有意義——橋掛掉時一樣綠燈——必須與同一時窗的正向 ping 配成一對才有證據力。
#include <algorithm>
#include <chrono>
#include <cstring>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/header.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

using namespace std::chrono_literals;
static const char *REPLY_TOPIC = "/goal_pose";

struct Opt {
  std::string role, mode = "small", label, topic;
  int count = 100, points = 1080, width = 640, height = 480;
  double rate = 10.0, seconds = 15.0;
};

static rclcpp::QoS qos() { return rclcpp::QoS(rclcpp::KeepLast(10)).reliable(); }

static std::string resolve(const Opt &o) {
  if (!o.topic.empty()) return o.topic;
  if (o.mode == "small") return "/initialpose";
  if (o.mode == "scan") return "/scan";
  return "/d455/color/image_raw";
}

class Pong : public rclcpp::Node {
public:
  Pong(const Opt &o) : Node("zpong_cpp") {
    pub_ = create_publisher<geometry_msgs::msg::PoseStamped>(REPLY_TOPIC, qos());
    auto reply = [this](const std_msgs::msg::Header &h) {
      geometry_msgs::msg::PoseStamped out;
      out.header.stamp = h.stamp;              // 抄回原 stamp = ping 端時鐘
      out.header.frame_id = "zpong_cpp";
      pub_->publish(out);
    };
    const std::string topic = resolve(o);
    if (o.mode == "small")
      sub_small_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
          topic, qos(),
          [reply](geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr m) { reply(m->header); });
    else if (o.mode == "scan")
      sub_scan_ = create_subscription<sensor_msgs::msg::LaserScan>(
          topic, qos(), [reply](sensor_msgs::msg::LaserScan::SharedPtr m) { reply(m->header); });
    else
      sub_img_ = create_subscription<sensor_msgs::msg::Image>(
          topic, qos(),
          [reply](sensor_msgs::msg::Image::SharedPtr m) { reply(m->header); });
  }

private:
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr sub_small_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_scan_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_img_;
};

class Ping : public rclcpp::Node {
public:
  std::vector<double> rtts;
  int sent = 0;

  Ping(const Opt &o) : Node("zping_cpp"), o_(o) {
    const std::string topic = resolve(o);
    if (o.mode == "small") pub_small_ = create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(topic, qos());
    else if (o.mode == "scan") pub_scan_ = create_publisher<sensor_msgs::msg::LaserScan>(topic, qos());
    else pub_img_ = create_publisher<sensor_msgs::msg::Image>(topic, qos());

    sub_ = create_subscription<geometry_msgs::msg::PoseStamped>(
        REPLY_TOPIC, qos(), [this](geometry_msgs::msg::PoseStamped::SharedPtr m) {
          int64_t sent_ns = static_cast<int64_t>(m->header.stamp.sec) * 1000000000LL + m->header.stamp.nanosec;
          rtts.push_back((now().nanoseconds() - sent_ns) / 1e6);
        });

    timer_ = create_wall_timer(std::chrono::duration<double>(1.0 / o.rate), [this]() { tick(); });
  }

private:
  void tick() {
    if (sent >= o_.count) return;
    auto stamp = now();
    if (o_.mode == "small") {
      geometry_msgs::msg::PoseWithCovarianceStamped m;
      m.header.stamp = stamp; m.header.frame_id = "zping";
      pub_small_->publish(m);
    } else if (o_.mode == "scan") {
      sensor_msgs::msg::LaserScan m;
      m.header.stamp = stamp; m.header.frame_id = "zping";
      m.angle_min = -3.14f; m.angle_max = 3.14f;
      m.angle_increment = 6.28f / o_.points;
      m.range_min = 0.1f; m.range_max = 30.0f;
      m.ranges.assign(o_.points, 1.0f);
      pub_scan_->publish(m);
    } else {
      sensor_msgs::msg::Image m;
      m.header.stamp = stamp; m.header.frame_id = "zping";
      m.width = o_.width; m.height = o_.height; m.encoding = "rgb8"; m.step = o_.width * 3;
      m.data.assign(static_cast<size_t>(o_.width) * o_.height * 3, 0);
      pub_img_->publish(m);
    }
    sent++;
  }

  Opt o_;
  rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pub_small_;
  rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr pub_scan_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_img_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

// 否定測試：訂閱一個「不該送達」的 topic，收到任何一則即為失敗。
// 刻意固定用 BEST_EFFORT 訂閱：BEST_EFFORT reader 同時相容 RELIABLE 與 BEST_EFFORT
// writer，若用 RELIABLE，QoS 不相容會讓訂閱根本配不上發布端，測試會假性通過。
class ExpectNone : public rclcpp::Node {
public:
  int n = 0;

  ExpectNone(const Opt &o) : Node("zexpect_none_cpp") {
    auto q = rclcpp::QoS(rclcpp::KeepLast(10)).best_effort();
    const std::string topic = resolve(o);
    auto count = [this](const std_msgs::msg::Header &) { n++; };
    if (o.mode == "small")
      sub_small_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
          topic, q,
          [count](geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr m) { count(m->header); });
    else if (o.mode == "scan")
      sub_scan_ = create_subscription<sensor_msgs::msg::LaserScan>(
          topic, q, [count](sensor_msgs::msg::LaserScan::SharedPtr m) { count(m->header); });
    else
      sub_img_ = create_subscription<sensor_msgs::msg::Image>(
          topic, q, [count](sensor_msgs::msg::Image::SharedPtr m) { count(m->header); });
  }

private:
  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr sub_small_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_scan_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_img_;
};

static double pct(std::vector<double> v, double p) {
  std::sort(v.begin(), v.end());
  size_t i = static_cast<size_t>(p / 100.0 * (v.size() - 1) + 0.5);
  return v[std::min(i, v.size() - 1)];
}

static const char *USAGE =
    "usage: zping --role {ping|pong|expect-none} [--mode {small|scan|image}] [--topic T]\n"
    "             [--count N] [--rate HZ] [--seconds S] [--points N] [--width W] [--height H]\n"
    "             [--label L]\n"
    "note: flood/sink 只有 zping.py 有\n";

// 未知的 --role 若沉默地落到 Ping，探針會變成發布端：small 模式會往 /initialpose
// 灌 100 則零位姿，那是白名單內的 topic，會過橋餵給車上的 AMCL。寧可直接拒絕啟動。
static bool parse(int argc, char **argv, Opt &o) {
  for (int i = 1; i < argc; i++) {
    std::string a = argv[i];
    if (a == "--ros-args") break;
    if (i + 1 >= argc) {
      fprintf(stderr, "zping: %s 缺少值\n%s", a.c_str(), USAGE);
      return false;
    }
    std::string v = argv[++i];
    if (a == "--role") o.role = v;
    else if (a == "--mode") o.mode = v;
    else if (a == "--label") o.label = v;
    else if (a == "--topic") o.topic = v;
    else if (a == "--count") o.count = std::stoi(v);
    else if (a == "--rate") o.rate = std::stod(v);
    else if (a == "--points") o.points = std::stoi(v);
    else if (a == "--width") o.width = std::stoi(v);
    else if (a == "--height") o.height = std::stoi(v);
    else if (a == "--seconds") o.seconds = std::stod(v);
    else {
      fprintf(stderr, "zping: 不認得的參數 %s\n%s", a.c_str(), USAGE);
      return false;
    }
  }
  if (o.role != "ping" && o.role != "pong" && o.role != "expect-none") {
    fprintf(stderr, "zping: 不認得的 --role '%s'\n%s", o.role.c_str(), USAGE);
    return false;
  }
  if (o.mode != "small" && o.mode != "scan" && o.mode != "image") {
    fprintf(stderr, "zping: 不認得的 --mode '%s'\n%s", o.mode.c_str(), USAGE);
    return false;
  }
  return true;
}

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  Opt o;
  if (!parse(argc, argv, o)) { rclcpp::shutdown(); return 2; }

  if (o.role == "pong") {
    rclcpp::spin(std::make_shared<Pong>(o));
    rclcpp::shutdown();
    return 0;
  }

  if (o.role == "expect-none") {
    auto node = std::make_shared<ExpectNone>(o);
    const std::string topic = resolve(o);
    auto end = std::chrono::steady_clock::now() +
               std::chrono::duration_cast<std::chrono::steady_clock::duration>(
                   std::chrono::duration<double>(o.seconds));
    rclcpp::executors::SingleThreadedExecutor exec;
    exec.add_node(node);
    while (rclcpp::ok() && std::chrono::steady_clock::now() < end) exec.spin_once(50ms);
    bool ok = node->n == 0;
    printf("RESULT {\"label\": \"%s\", \"impl\": \"cpp\", \"role\": \"expect-none\", "
           "\"mode\": \"%s\", \"topic\": \"%s\", \"seconds\": %.2f, \"recv\": %d, "
           "\"pass\": %s}\n",
           o.label.c_str(), o.mode.c_str(), topic.c_str(), o.seconds, node->n,
           ok ? "true" : "false");
    if (ok) printf("PASS: no messages on %s in %.1fs\n", topic.c_str(), o.seconds);
    else printf("FAIL: received %d message(s) on %s in %.1fs\n", node->n, topic.c_str(), o.seconds);
    rclcpp::shutdown();
    return ok ? 0 : 1;
  }

  auto node = std::make_shared<Ping>(o);
  auto deadline = std::chrono::steady_clock::now() +
                  std::chrono::duration_cast<std::chrono::steady_clock::duration>(
                      std::chrono::duration<double>(o.count / o.rate + 8.0));
  // 必須是帶 timeout 的 spin_once：無事可做時 spin_some 會立刻返回，這個迴圈就把一個
  // 核心跑滿，而 CPU 競爭會灌水這支探針自己要量的 RTT（Python 版同樣用 0.05s）。
  rclcpp::executors::SingleThreadedExecutor exec;
  exec.add_node(node);
  while (rclcpp::ok() && std::chrono::steady_clock::now() < deadline &&
         static_cast<int>(node->rtts.size()) < o.count)
    exec.spin_once(50ms);

  auto &r = node->rtts;
  printf("RESULT {\"label\": \"%s\", \"impl\": \"cpp\", \"mode\": \"%s\", \"sent\": %d, \"recv\": %zu",
         o.label.c_str(), o.mode.c_str(), node->sent, r.size());
  if (!r.empty())
    printf(", \"rtt_min_ms\": %.3f, \"rtt_p50_ms\": %.3f, \"rtt_p90_ms\": %.3f, \"rtt_p99_ms\": %.3f, \"rtt_max_ms\": %.3f",
           *std::min_element(r.begin(), r.end()), pct(r, 50), pct(r, 90), pct(r, 99),
           *std::max_element(r.begin(), r.end()));
  printf("}\n");
  rclcpp::shutdown();
  return 0;
}
